mmaction/datasets/tube_dataset.py

```diff
@@ -0,0 +1,183 @@
+        ])
+
+    def check_tubelet(self, video_tubes, frame_index, tube_length):
+        is_whole = all([
+            self.tubelet_in_tube(video_tube, frame_index, tube_length)
+            or self.tubelet_out_tube(video_tube, frame_index, tube_length)
+            for video_tube in video_tubes
+        ])
+        has_gt = any([
+            self.tubelet_in_tube(tube, frame_index, tube_length)
+            for tube in video_tubes
+        ])
+        return is_whole and has_gt
+
+    def load_annotations(self):
+        pkl_data = pickle.load(
+            open(self.ann_file, 'rb'), encoding=self.encoding)
+        gt_tubes = pkl_data['gttubes']
+        num_frames = pkl_data['nframes']
+        train_videos = pkl_data['train_videos']
+        test_videos = pkl_data['test_videos']
+        resolution = pkl_data['resolution']
+
+        self.gt_tubes = gt_tubes
+        self.labels = pkl_data['labels']
+
+        assert len(train_videos[self.split - 1]) + len(
+            test_videos[self.split - 1]) == len(num_frames)
+        videos = train_videos[
+            self.split - 1] if not self.test_mode else test_videos[self.split -
+                                                                   1]
+        self.videos = videos
+
+        if self.preload_ann_file is not None and osp.exists(
+                self.preload_ann_file):
+            video_infos = pickle.load(open(self.preload_ann_file, 'rb'))
+        else:
+            video_infos = []
+
+            for video in videos:
+                video_tubes = sum(gt_tubes[video].values(), [])
+                for i in range(1, num_frames[video] + 2 - self.tube_length):
+                    if self.check_tubelet(video_tubes, i, self.tube_length):
+                        frame_dir = video
+                        if self.data_prefix is not None:
+                            frame_dir = osp.join(self.data_prefix, video)
+                        gt_bboxes = defaultdict(list)
+
+                        for label_index, tubes in gt_tubes[video].items():
+                            for tube in tubes:
+                                if i not in tube[:, 0]:
+                                    continue
+                                boxes = tube[
+                                    (tube[:, 0] >= i) *
+                                    (tube[:, 0] < i + self.tube_length), 1:5]
+                                gt_bboxes[label_index].append(boxes)
+                        video_info = {}
+                        video_info['indice'] = (video, i)
+                        video_info['video'] = video
+                        video_info['frame_dir'] = frame_dir
+                        video_info['total_frames'] = num_frames[video]
+                        video_info['resolution'] = resolution[video]
+                        video_info['gt_bboxes'] = gt_bboxes
+                        video_infos.append(video_info)
+
+            if self.save_preload:
```

Comment:
Seems different tube length may cause some difference to the videos infos, can we do the other parts of work during preprocessing, and let `load_annotation` focus on processing with different tube_length?